### PR TITLE
Switch auth to go against actual http server. Add service unit tests.

### PR DESCRIPTION
**Pull request by BrendanFDMoore**
This branch adds proper unit testing to the stack, with full coverage for the newly created `AuthenticationService` and `ServerService` files.

Authentication now hits the `ngcourse` [server](https://ngcourse.herokuapp.com/auth/login) (via local proxy for CORS circumvention) which allows *any* credentials through.

Connected to rangle/rangle-starter#91
Connected to https://github.com/rangle/rangle-starter/issues/59

### CODE REVIEW

**Comment by BrendanFDMoore** [comment]
This is now a stable/tidied version with http auth requests handled by the node server against a json file. @codebyuma and I are still working on getting a passport-based version working, but did not have success on Friday.

File: src/services/authentication/index.test.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import {AuthenticationService} from './';
+
+describe('AuthenticationService', () => {
+  let _mockServerService;
+  let _mockCreds;
+  let _mockResponse;
+
+  beforeEach(() => {
+    _mockCreds = { username: 'alice', password: 'x'};
+    _mockResponse = {
```
Comment (SethDavenport) [inline src/services/authentication/index.test.ts]: This response is only used in one of the tests; I would be inclined to initialize it in that test instead of in `beforeEach`.

File: src/services/authentication/index.test.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    
+    expect(_mockServerService).to.not.be.undefined;
+    let authService = new AuthenticationService(_mockServerService);
+    
+    return authService.login(_mockCreds)
+      .then(data => {
+        // Assertions thrown here will result to a failed promise downstream.
+        expect(data).to.deep.equal(_mockResponse);
+        // Remember to call done(), otherwise the test will time out (and fail).
+        done();
```
Comment (SethDavenport) [inline src/services/authentication/index.test.ts]: These calls to `done()` should not be needed - mocha supports async tests that return a promise.  So instead,

```typescript
it('should receive successful response', () => {    
    expect(_mockServerService).to.not.be.undefined;
    let authService = new AuthenticationService(_mockServerService);
    
    return authService.login(_mockCreds)
      .then(data => {
        // Assertions thrown here will result to a failed promise downstream.
        expect(data).to.deep.equal(_mockResponse);
      });
  });
```

Note that I removed the done parameter; mocha will instead detect that this test returns a promise and wait for it to resolve.

Also note that I got rid of the error block; if the promise is rejected mocha will fail the test for us.

File: src/services/server/index.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
+import Inject from '../../utils/di';
+
+export class ServerService {
+
+  static $inject = [
+    '$http'
+  ];
+
+  // Actual server URL to be accessed, but blocked due to CORS
+  // private BASE_URL = 'http://ngcourse.herokuapp.com';
```
Comment (SethDavenport) [inline src/services/server/index.ts]: This isn't true any more.

File: server/auth.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
+var fs = require("fs");
+var util = require('util');
+
+module.exports = (app) => {
```
Comment (SethDavenport) [inline server/auth.js]: We decided not to go with passport in the end? That's OK for this test project but it might be worth putting in a comment to use 'read' auth solutions in production code...

File: server/auth.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  var userFileData = fs.readFileSync("./server/users.json");
```
Comment (SethDavenport) [inline server/auth.js]: Should probably avoid readFileSync in a webserver, for performance reasons.

You can do 

```javascript
fs.readFile('./server/users.json', (err, userFileData) => {
  if (err) {
    // Tell the browser something's wrong so it closes it's connection.
    res.status(500).send('Internal server error');

    // log the error for debugging (but don't send it to the client, for security).
    winston.error(err);

    // re-throw the error so the server dies and can be restarted by whatever deployment
    // manager we're using (counterintuitive, but a NodeJS best practice: https://www.joyent.com/developers/node/design/errors)
    throw err; 
  }

  // etc.
  res.status(200).send(etc);
});

Comment (SethDavenport) [inline server/auth.js]: Alternately (and I tend to prefer this) you can use `denodeify` to get a promise-based version of fs.ReadFile:

```javascript
const readFile = denodeify(fs.readFile);

// ...

readFile('./server/users.json')
  .then(userFileData => {
    // etc.
    res.status(200).send(etc);
  })
  .catch(err => {
    // Tell the browser something's wrong so it closes its connection.
    res.status(500).send('Internal server error');

    // log the error for debugging (but don't send it to the client, for security).
    winston.error(err);

    // re-throw the error so the server dies and can be restarted by whatever deployment
    // manager we're using (counterintuitive, but a NodeJS best practice: https://www.joyent.com/developers/node/design/errors)
    throw err; 
  });
```

File: server/auth.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    users.map(u => console.log(u.Username));
+
+    if (req.body.username && req.body.password){
+      const filtered = users.filter(
+        function(user){
+          return (user.Username===req.body.username) 
+          && (user.Password===req.body.password);
+        });
+      if (filtered.length>0)
+        res.status(200).send(JSON.stringify({data:{msg:"LOGIN SUCCESSFUL"},meta:{token:"abcd1234",expires:"2020-01-01"}}));
```
Comment (SethDavenport) [inline server/auth.js]: Single quotes please - I'l surprised the linter didn't complain.

File: server/proxy-config.js [hunk 7]
```diff
@@ -14,10 +14,10 @@
 module.exports = {
   // Calls to /api/foo will get routed to
   // http://jsonplaceholder.typicode.com/foo.
-  /*
+  
   '/api/': {
-    target: 'http://jsonplaceholder.typicode.com',
+    target: 'https://ngcourse.herokuapp.com/',
```
Comment (SethDavenport) [inline server/proxy-config.js]: Is this still needed?

File: server/auth.js (hunk 4 above)
Comment (BrendanFDMoore) [inline server/auth.js]: We haven't abandoned Passport, just didn't get it working by end of day Friday. We had this version going, so I just cleaned it up as a good enough example.

**Comment by BrendanFDMoore** [comment]
Issues have been resolved. Passport version will be revisited today.

**Comment by BrendanFDMoore** [comment]
@SethDavenport this is now done with passport integration. (Hi5 @codebyuma ). After we get a squirrel, I'll rebase to fix the minting and merge conflicts, unless you want that resolved beforehand.

File: src/simple.test.ts [hunk 8]
```diff
@@ -1,13 +1,18 @@
-describe('Simple Test', () => {
+'use strict';
```
Comment (SethDavenport) [inline src/simple.test.ts]: We can probably get rid of this file now that we have other, real tests.

File: server/auth.js [hunk 9]
```diff
@@ -0,0 +1,50 @@
+'use strict';
```
Comment (SethDavenport) [inline server/auth.js]: Is this file still needed now that we have passport?

**Comment by SethDavenport** [comment]
Otherwise, :shipit:
